Components: Add tests for Button component

nitrajka opened the pull request:
Hi, I would like to solve #7936 and write tests for `<Button/>` component. 

We need to simulate the rendering of the component and then test every possible variation of  `Button`'s styles.

Firstly, I was not sure, whether to use shallow rendering, mount or full DOM rendering. Then I realized, that shallow rendering is needed only when I don't want to render children and test their styles or behavior. 
So in my PR I use `shallow` when passing a text to the `Button` as a child. And `mount`, when I need to test, whether the child - `Gridicon` component has expected className.

I was also thinking about all possible variations of `Button`s and I found them in <a href="https://wpcalypso.wordpress.com/devdocs/design">DevDocs</a>. 
<img width="1280" alt="screen shot 2016-10-14 at 14 41 24" src="https://cloud.githubusercontent.com/assets/14272775/19387718/4d81a9be-921c-11e6-960b-7fd52f81085f.png">

Of course, I also need to test the button with `href` prop and `compact` prop set to true. But before I add other tests, I wanted to ask,
Is this good way testing it?

@aduth

aduth commented:
Hey @nitrajka , thanks for jumping on #7936!

Yes, ideally we should avoid `mount` and use `shallow` as much as possible, since it doesn't require a full DOM to be instantiated and helps isolate the component under test. The "downside" that it doesn't provide access to children components is really not a downside in that those components should already be tested separately anyways (themselves also tested with `shallow`, so on and so on).

You can read more about the reasoning behind shallow rendering here:

- https://facebook.github.io/react/docs/test-utils.html#shallow-rendering
- https://github.com/airbnb/enzyme/blob/master/docs/api/shallow.md

I've not yet taken a close look at the code, but I do see `mount` used a few times. Is there any way we could avoid that? What is it about the test case that requires a full mount?

nitrajka commented:
Alright then. I'll try to test it with `shallow` only.

I just thought myself, that we might need to test also the child's (`Gridicon`'s) `className` as we send it as a prop of that child  (`Gridicon`). Therefore I used `mount` instead of `shallow`. But you are right. We don't need to test the behaviour of `Gridicon`, because it was already tested.

Of course, there is a way to avoid `mount`. Thank you for your idea. I'll fix it soon.

aduth commented:
You're right that with `shallow` you can't test the `className` of the rendered `<Gridicon />` result, but you _can_ (and should) test that the prop passed to the `<Gridicon />` is the one that's expected.

I expect it'll look something like:

```js
expect( button.find( Gridicon ) ).to.have.prop( 'icon' ).equal( 'arrow-left' );
```

nitrajka commented:
I see now. That's a good idea! Thank you for an advice.

nitrajka commented:
I wrote the tests for `<a>` Button which differs from the normal `<Button>` only in `href` property. Also added tests for compact `Button`s. I considered all possible variations of the `Button` component.

When writing a test (for instance in row 34), should I use `'.button'` or `'button'` CSS selector? I know, that in `Calypso`'s CSS files, you don't use element selectors but only class selectors. Is it the same in tests?

I know, that `<a>` tag can't be `disabled` - so I'm going to delete the describe section which tests that...

Besides that, what do you think about tests, which I've written? @aduth 

To test it run: `npm run test-client client/components/button/test/index.js`
All of the tests are passing.

aduth commented:
>When writing a test (for instance in row 34), should I use `'.button'` or `'button'` CSS selector? I know, that in `Calypso`'s CSS files, you don't use element selectors but only class selectors. Is it the same in tests?

I don't think this matters too much, as long as you know that it'll match the intended target, or if you want to ensure that a particular class or DOM node is rendered (might be a good idea for a base test).

I'll plan to look over the changes shortly and submit a review.

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+/**
+ * Created by nitrajka on 09/10/16.
```
Let's leave out comments like this, since it's better captured and maintained by git history.

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+	describe( 'not compact', function() {
+		describe( 'button tag', function() {
+			const text = 'Text';
+			const icon = ( <Gridicon size={ 18 } icon="arrow-left" /> );
```
Fine to leave them, but as an FYI the parentheses aren't necessary.

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+				} );
+
+				it( 'renders disabled', function() {
+					const button = shallow( <Button disabled>{ disabled }</Button> );
```
Is there value in assigning the `disabled` text to a variable versus simply inlining it, i.e.

```jsx
<Button disabled>Disabled</Button>
```

Do we need children at all?

```jsx
<Button disabled />
```

aduth replied:
Also, to the naming of the test case, it's not necessarily testing that the component can be disabled, but rather that it passes all props aside from its own through to rendered tag.

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+				it( 'renders with text', function() {
+					const button = shallow( <Button>{ text }</Button> );
+					expect( button ).to.contain( 'Text' );
+					expect( button.is( '.button' ) ).to.equal( true );
```
`chai-enzyme` gives us a [`className` assertion](https://github.com/producthunt/chai-enzyme#classnamestr), which would read more clearly:

```js
expect( button ).to.have.className( 'button' );
```

As an aside, Chai has a [`.true` assertion](http://chaijs.com/api/bdd/#method_true) too which could be used in place of yours here:

```js
expect( button.is( '.button' ) ).to.be.true;
```

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+import Button from '../';
+import Gridicon from 'components/gridicon';
+
+describe( 'Button', function() {
```
With ES6, you could use arrow functions here and throughout as a more concise alternative to the anonymous function, especially when you don't need to access `this`, which is generally true in our tests.

```js
describe( 'Button', () => {
```

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+			describe( 'basic styled', function() {
+				it( 'renders with text', function() {
+					const button = shallow( <Button>{ text }</Button> );
+					expect( button ).to.contain( 'Text' );
```
Again, I'm not sure that assigning the string to a variable adds much value, but if you were to keep it, I'd make sure that the `contain` assertion here checks against the variable, not the text itself (e.g. what if the variable were to change? we'd expect all tests to still pass).

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+					expect( button.is( '.button' ) ).to.equal( true );
+				} );
+
+				it( 'renders with text and icon', function() {
```
I don't know that there's any benefit in this test beyond what's already achieved in the basic render test. Since the `<Button />` component simply renders the children passed to it, we should be reasonably confident that all of the props of descendent components are unaffected.

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+					it( 'renders with text', function() {
+						const button = shallow( <Button scary={ true }>{ text }</Button> );
+						expect( button ).to.contain( 'Text' );
+						expect( button.is( '.button.is-scary' ) ).to.equal( true );
```
Could be two separate `have.className` assertions?

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+
+				it( 'renders disabled', function() {
+					const button = shallow( <Button disabled>{ disabled }</Button> );
+					expect( button.find( '.button' ) ).to.be.disabled();
```
I think calling the `disabled` as a function, even if it is, may not be necessary? At least consistent with how we typically call assertions:

```js
expect( button.find( '.button' ) ).to.be.disabled;
```

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+
+					it( 'renders with icon', function() {
+						const button = shallow(
+							<Button scary={ true }>
```
As an FYI, you can leave out `={ true }`, so just `<Button scary />`

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+						expect( button.is( '.button.is-scary' ) ).to.equal( true );
+					} );
+
+					it( 'renders with text and icon', function() {
```
Again, I don't know that we need more than one test for checking basic render behavior.

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+					expect( button.is( '.button.is-primary' ) ).to.equal( true );
+				} );
+
+				it( 'renders with text and icon', function() {
```
As above, I don't think we need more than one test for checking basic render behavior.

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,1064 @@
+					expect( button.is( '.button.is-borderless' ) ).to.equal( true );
+				} );
+
+				it( 'renders with text and icon', function() {
```
As above, I don't think we need more than one test for checking basic render behavior.

nitrajka commented:
I'm ready for next review. :)

nitrajka commented:
Yes, we can. Done :) any other suggestions?

aduth commented on client/components/button/test/index.js:
```diff
@@ -0,0 +1,89 @@
+
+			it( '`type` prop changes when overridden', () => {
+				const typeProp = 'submit';
+				const button = shallow( <Button target={ blank } rel={ relString } type ={ typeProp } /> );
```
Minor: Stray space between `type` and `=` should be removed.

nitrajka commented:
To the first point, you've suggested:
I wanted to test that **the expected** element (`<a>` or `<button>`) has/has not some exact property. Therefore I added `find( 'a' )` or `find( 'button' )`. 
For example: If you want to render `<a>` and test whether it has/has not `type` prop, an `<a>` element has to be found instead `<button>`. But okay, I did it as you suggested.

aduth commented:
@nitrajka Ah, I see what you mean. Yeah, I'd agree it's a good idea to test that it's the expected element that's being rendered. We could have a separate assertion like:

```js
expect( button ).to.match( 'a' );
```

Sorry, I'd add this myself but since it's an external repository I'm not able to make the changes. I think the tests are looking very good, so I'd be okay to merge now with or without this change. What do you think?

nitrajka commented:
I'd like to add this assertion before you merge it. Plese, don't do it yet. 
Now I don't have much time as I'm attendee of Reactive Conf in Bratislava. But soon it'll be done. Will ping you then.

aduth commented:
>I'm attendee of Reactive Conf in Bratislava

Enjoy the conference 😄

nitrajka commented:
I rewrote the descriptions of tests because I found those previous not as easy to understand. Also added the assertion which tests whether the rendered component is the expected one. Tests are passing  :) @aduth

nitrajka commented:
I'm really glad I could help! 😃 Thank you for your reviews, time and suggestions :)
